rclcpp/src/rclcpp/node_interfaces/node_parameters.cpp

```diff
@@ -806,14 +809,20 @@ NodeParameters::get_parameters(const std::vector<std::string> & names) const
 rclcpp::Parameter
 NodeParameters::get_parameter(const std::string & name) const
 {
-  rclcpp::Parameter parameter;
+  std::lock_guard<std::recursive_mutex> lock(mutex_);
 
-  if (get_parameter(name, parameter)) {
-    return parameter;
+  auto param_iter = parameters_.find(name);
+  if (
+    parameters_.end() != param_iter &&
+    param_iter->second.value.get_type() != rclcpp::ParameterType::PARAMETER_NOT_SET)
+  {
+    return rclcpp::Parameter{name, param_iter->second.value};
   } else if (this->allow_undeclared_) {
-    return parameter;
-  } else {
+    return rclcpp::Parameter{};
+  } else if (parameters_.end() == param_iter) {
     throw rclcpp::exceptions::ParameterNotDeclaredException(name);
+  } else {
+    throw rclcpp::exceptions::NoParameterOverrideProvided(name);
```

Comment:
This should maybe be `ParameterUninitializedException`.
It's not exactly `NoParameterOverrideProvided`, because this is cause by an override not being provided and the user not setting the parameter before getting it.

If we change this, `NoParameterOverrideProvided` wouldn't be thrown anymore and should be deprecated.

Reply:
If we want to try and get this in Galactic, do you think it is okay to deprecate it? I can add the new exception in any case.

Reply:
hmm.. `NoParameterOverrideProvided` was only added for the dynamic v. static type feature. If we're willing to change API just before the Galactic release then we can avoid a deprecation cycle :sweat_smile:

Reply:
> hmm.. NoParameterOverrideProvided was only added for the dynamic v. static type feature. If we're willing to change API just before the Galactic release then we can avoid a deprecation cycle sweat_smile

:+1:

Reply:
Added `ParameterUninitializedException` (0d97e32).
Removed `NoParameterOverrideProvided` (b32f19d). I'll check for any downstream usage.